### PR TITLE
adding block volume support for tech preview

### PR DESCRIPTION
**Pull request by screeley44**
based on alpha block volume support in upstream kube version 1.9

### CODE REVIEW

**Comment by screeley44** [comment]
cc @erinboyd  - not sure who else to copy any more, I know our docs contacts have changed, can you copy anyone else that is needed for this?

**Comment by ahardin-rh** [comment]
@screeley44 Thanks! Is this work associated with a Trello card? Also, I assume this is targeting OCP 3.9?

If you're ever blocked by docs, you can ping our documentation program manager @vikram-redhat to get things moving and assigned to a writer. Since this is storage related, this would likely get assigned to @tmorriso-rh

File: architecture/additional_concepts/storage.adoc [hunk 1]
```diff
@@ -578,3 +578,175 @@ spec:
...
+feedback during the development process.
+
+For more information on Red Hat Technology Preview features support scope, see
+https://access.redhat.com/support/offerings/techpreview/.
+endif::[]
+====
+
+You can statically provisioning Raw Block Volumes by including some new API fields in your PV and PVC specs.
+
+Example Persistent Volume:
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/Example Persistent Volume:/.Example Persistent Volume

File: architecture/additional_concepts/storage.adoc [hunk 2]
```diff
@@ -578,3 +578,175 @@ spec:
...
+  accessModes:
+    - ReadWriteOnce
+  volumeMode: Block <1>
+  persistentVolumeReclaimPolicy: Retain
+  fc:
+    targetWWNs: ["50060e801049cfd1"]
+    lun: 0
+    readOnly: false
+----
+<1> volumeMode field indicating that this PV is a raw block volume
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: `volumeMode` field indicating that this PV is a raw block volume.

File: architecture/additional_concepts/storage.adoc [hunk 3]
```diff
@@ -578,3 +578,175 @@ spec:
...
+not recommend to use them for production. These features provide early access to
+upcoming product features, enabling customers to test functionality and provide
+feedback during the development process.
+
+For more information on Red Hat Technology Preview features support scope, see
+https://access.redhat.com/support/offerings/techpreview/.
+endif::[]
+====
+
+You can statically provisioning Raw Block Volumes by including some new API fields in your PV and PVC specs.
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/provisioning/provision
s/Raw Block Volumes/raw block volumes
s/specs/specifications

File: architecture/additional_concepts/storage.adoc [hunk 4]
```diff
@@ -578,3 +578,175 @@ spec:
...
+spec:
+  accessModes:
+    - ReadWriteOnce
+  volumeMode: Block <1>
+  resources:
+    requests:
+      storage: 10Gi
+----
+<1> VolumeMode field indicating that a raw block Persistent Volume is requested.
+
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: `VolumeMode` field indicating that a raw block persistent volume is requested.

File: architecture/additional_concepts/storage.adoc [hunk 5]
```diff
@@ -578,3 +578,175 @@ spec:
...
+  volumeMode: Block <1>
+  persistentVolumeReclaimPolicy: Retain
+  fc:
+    targetWWNs: ["50060e801049cfd1"]
+    lun: 0
+    readOnly: false
+----
+<1> volumeMode field indicating that this PV is a raw block volume
+
+Example Persistent Volume Claim:
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/Example Persistent Volume Claim:/.Example Persistent Volume Claim

File: architecture/additional_concepts/storage.adoc [hunk 6]
```diff
@@ -578,3 +578,175 @@ spec:
...
+|Filesystem
+|No Bind
+|===
+
+Status of Upstream Plugins that do or will support Block Volumes:
+
+[cols="1,2",options="header"]
+|===
+
+|Plugin |Support Block Volume
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/Plugin/Plug-in

File: architecture/additional_concepts/storage.adoc [hunk 7]
```diff
@@ -578,3 +578,175 @@ spec:
...
+    - ReadWriteOnce
+  volumeMode: Block <1>
+  resources:
+    requests:
+      storage: 10Gi
+----
+<1> VolumeMode field indicating that a raw block Persistent Volume is requested.
+
+Example Pod Spec:
+[source, yaml]
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/Example Pod Spec:/.Example Pod Specification

File: architecture/additional_concepts/storage.adoc [hunk 8]
```diff
@@ -578,3 +578,175 @@ spec:
...
+      args: [ "tail -f /dev/null" ]
+      volumeDevices:  <1>
+        - name: data
+          devicePath: /dev/xvda <2>
+  volumes:
+    - name: data
+      persistentVolumeClaim:
+        claimName: block-pvc <3>
+----
+<1> VolumeDevices (similar to VolumeMounts) is used for Block Devices and can only be used with PersistentVolumeClaim sources
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: `VolumeDevices` (similar to `VolumeMounts`) is used for block devices and can only be used with `PersistentVolumeClaim` sources.

File: architecture/additional_concepts/storage.adoc [hunk 9]
```diff
@@ -578,3 +578,175 @@ spec:
...
+<2> DevicePath (similar to mountPath) represents the path to the physical device
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: `DevicePath` (similar to `mountPath`) represents the path to the physical device.

File: architecture/additional_concepts/storage.adoc [hunk 10]
```diff
@@ -578,3 +578,175 @@ spec:
...
+<3> The volume source needs to be of type persistentVolumeClaim and should match the name of the PVC as expected
+
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: The volume source must be of type `persistentVolumeClaim` and must match the name of the PVC as expected.

File: architecture/additional_concepts/storage.adoc [hunk 11]
```diff
@@ -578,3 +578,175 @@ spec:
...
+
+Accepted Values for VolumeMode:
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: s/Accepted Values for VolumeMode:/.Accepted Values for VolumeMode

File: architecture/additional_concepts/storage.adoc [hunk 12]
```diff
@@ -578,3 +578,175 @@ spec:
...
+|Yes
+
+|Block
+|No
+|===
+
+Binding Scenarios for Block Volumes:
+[IMPORTANT]
+====
+Unspecified results in default value of *Filesystem*
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: Unspecified values results in the default value of `Filesystem`.

File: architecture/additional_concepts/storage.adoc [hunk 13]
```diff
@@ -578,3 +578,175 @@ spec:
         claimName: myclaim
 
 ----
+
+[[blocl-volume-support]]
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: [[block-volume-support]]

File: architecture/additional_concepts/storage.adoc [hunk 14]
```diff
@@ -578,3 +578,175 @@ spec:
...
+|Filesystem
+|Block
+|No Bind
+
+|Block
+|Filesystem
+|No Bind
+|===
+
+Status of Upstream Plugins that do or will support Block Volumes:
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: .Status of Upstream Plug-ins That Support or Will Support Block Volumes

File: architecture/additional_concepts/storage.adoc [hunk 15]
```diff
@@ -578,3 +578,175 @@ spec:
...
+|Value |Default
+
+|Filesystem
+|Yes
+
+|Block
+|No
+|===
+
+Binding Scenarios for Block Volumes:
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: Binding Scenarios for Block Volumes

File: architecture/additional_concepts/storage.adoc (hunk 4 above)
Comment (erinboyd) [reply architecture/additional_concepts/storage.adoc]: @screeley44 lets keep the casing for volumeMode consistent to match the API

**Comment by vikram-redhat** [comment]
Thanks @ahardin-rh and @screeley44 . Yes, @tmorriso-rh is the writer working with the storage team and attends regular calls. Please make sure to bring it to her attention in those calls or by tagging her.

**Comment by screeley44** [comment]
thanks @ahardin-rh , @erinboyd , @vikram-redhat, @tmorriso-rh - I think I addressed all the fixes, PTAL

**Comment by screeley44** [comment]
@ahardin-rh - here is the trello link   

https://trello.com/c/rNFC3blm/547-13-implement-raw-block-storage-design-proposal

File: architecture/additional_concepts/storage.adoc [hunk 16]
```diff
@@ -578,3 +578,175 @@ spec:
...
+  volumeMode: Block <1>
+  persistentVolumeReclaimPolicy: Retain
+  fc:
+    targetWWNs: ["50060e801049cfd1"]
+    lun: 0
+    readOnly: false
+----
+<1> `volumeMode` field indicating that this PV is a raw block volume.
+
+Example Persistent Volume Claim
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: .Example Persistent Volume Claim

File: architecture/additional_concepts/storage.adoc [hunk 17]
```diff
@@ -578,3 +578,175 @@ spec:
...
+  accessModes:
+    - ReadWriteOnce
+  volumeMode: Block <1>
+  resources:
+    requests:
+      storage: 10Gi
+----
+<1> `volumeMode` field indicating that a raw block persistent volume is requested.
+
+Example Pod Specification
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: .Example Pod Specification

File: architecture/additional_concepts/storage.adoc [hunk 18]
```diff
@@ -578,3 +578,175 @@ spec:
...
+feedback during the development process.
+
+For more information on Red Hat Technology Preview features support scope, see
+https://access.redhat.com/support/offerings/techpreview/.
+endif::[]
+====
+
+You can statically provision raw block volumes by including some new API fields in your PV and PVC specifications.
+
+Example Persistent Volume
```
Comment (ahardin-rh) [inline architecture/additional_concepts/storage.adoc]: .Example Persistent Volume

**Comment by ahardin-rh** [comment]
@screeley44 Thanks! Overall, LGTM. I just made one formatting nit that applies to the 3 examples. Otherwise, I think this is ready to go. 

Also, thanks for the Trello link. As a heads-up, if you apply the documentation label to the Trello card when it's in the "New" state on your board, then it will clone to our documentation team's Trello board/content plan. Then, @vikram-redhat can assign a writer to the card. Thanks!

**Comment by screeley44** [comment]
fixed the nits - let me know if anything else is needed...THANKS!!

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../architecture/additional_concepts/storage.adoc#architecture-additional-concepts-storage[Additional Concepts -> Persistent Storage]
|Added a new section on xref:../architecture/additional_concepts/storage.adoc#block-volume-support[Block Volume Support].
%

**Comment by vikram-redhat** [comment]
@jhou - could we get a QE ack for this feature?

**Comment by jhou** [comment]
I think you may have tagged the wrong person.

**Comment by ahardin-rh** [comment]
Whoops! Sorry, Jeremy! 
@jhou1 Can you please give this a QE review? Thanks!

**Comment by jhou1** [comment]
@qinpingli PTAL, thanks

File: architecture/additional_concepts/storage.adoc [hunk 19]
```diff
@@ -578,3 +578,175 @@ spec:
...
+|Block
+|Filesystem
+|No Bind
+|===
+[IMPORTANT]
+====
+Unspecified values result in the default value of *Filesystem*.
+====
+
+.Status of Upstream Plug-ins That Support or Will Support Block Volumes
```
Comment (qinpingli) [inline architecture/additional_concepts/storage.adoc]: How about replace "the status of Upstream" with "the status of Openshift".

**Comment by qinpingli** [comment]
Add how to enable feature gate "BlockVolume=true" to use this feature?

**Pull request merged**